File: WordSearch.java
/**
 * LeetCode 79 Word Search
 */
class Solution {
    public boolean exist(char[][] board, String word) {
        if (board == null || word == null || word.length() == 0) {
            return false;
        }
        int maxH = board.length;
        int maxD = board[0].length;
        char startWith = word.charAt(0);
        int[][] direction = {{-1, 0}, {0, -1}, {0, 1}, {1, 0}};
        boolean[][] marked = new boolean[maxH][maxD];
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (board[i][j] == startWith) {
                    //find
                    if (backtrack(i,j,word,0,board,maxH,maxD,marked,direction)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    public boolean backtrack(int i, int j,String word,int cur, char[][] board ,int maxH, int maxD,boolean[][] marked,int[][] direction) {
        if (i >= maxH || i < 0 || j >= maxD || j < 0) {
            return false;
        }

        if (cur == word.length() - 1) {
            return board[i][j] == word.charAt(cur);
        }

        if (board[i][j] == word.charAt(cur)) {
            marked[i][j] = true;
            for (int k = 0; k < direction.length; k++) {
                int nextI = direction[k][0] + i;
                int nextJ = direction[k][1] + j;
                if (!(nextI >= maxH || nextI < 0 || nextJ >= maxD || nextJ < 0) && !marked[nextI][nextJ]) {
                    if (backtrack(nextI, nextJ, word, cur + 1, board, maxH, maxD, marked, direction)) {
                        return true;
                    }
                }
            }
            marked[i][j] = false;
        }
        return false;
    }
}

File: WordSearch.go
var dire = []int{-1,0,1,0,-1}


func exist(board [][]byte, word string) bool {

	k := len(word)

	h := len(board)
	w := len(board[0])

	res := false

	visited := make([][]bool,h)

	for i := range visited {
		visited[i] = make([]bool,w)
	}


	var backtrack func(int,int,int)
	backtrack = func(i int,j int,pos int) {

		if i < 0 || j < 0 || i >= h || j >= w {
			return
		}

		if visited[i][j] || res || board[i][j] != word[pos] {
			return
		}

		if pos == k - 1 {
			res = true
			return
		}
		visited[i][j] = true
		for x := 0;x < 4;x++ {
			ni := i + dire[x]
            nj := j + dire[x + 1]
		 	backtrack(ni,nj,pos + 1)
		}
		visited[i][j] = false
	}

	for i := 0;i < h;i++ {
		for j := 0;j < w;j++ {
			backtrack(i,j,0)
		}
	}
	return res
}
